### PR TITLE
Small cleanup when dataloader states are saved

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Resolves a small TODO in the code.
Includes #11842 which should be merged first.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [n/a] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 1]
```diff
@@ -165,20 +163,16 @@ def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
 
         if (
-            self.trainer is None
-            or not self.trainer.state._fault_tolerant_mode.is_enabled
-            or self._data_fetcher is None
-            or self._num_completed_batches_reached()  # did not finish
-            # TODO: fault-tolerance requires a minimum number of batches so probably should be > 0
-            or self.batch_progress.current.ready == 0  # did not start
+            self.trainer is not None
```
Comment (Borda) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:166]: it is just making it inverse?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:166]: No, it also got rid of

`        # TODO: this should use `pytorch_lightning/trainer/supporters.py::CombinedLoader._state_dict_fn``
